[BACKPORT] Fix dependencies [v0.13]

mrcasals opened the pull request:
#### :tophat: What? Why?
Ports #3732 to `0.13-stable`.

Partially ports #3756 to `0.13-stable`: blocks the `sassc` version, and adds the missing `bootsnap` to `decidim-generators`.

#### :pushpin: Related Issues
- Related to #3756, #3775.

mrcasals commented:
@deivid-rodriguez any idea why the generators suite is using an incorrect decidim version?

```
         In Gemfile:
           decidim was resolved to 0.14.0.dev, which depends on
             decidim-generators (~> 0.14.a)
```

https://circleci.com/gh/decidim/decidim/124721

mrcasals commented:
Mh, I don't get why the dependencies can't be met though 😕

deivid-rodriguez commented:
Yeah, indeed. Bundler should be able to resolve the situation but it's not... I run into this in the previous backport PR but couldn't find a solution. Then when you tried backporting it to 0.12 it worked, and now it's back in the 0.13 backport. One idea that comes to mind is that decidim can't handle the `.a` prelease specificator in this situation, so going back to "exact internal dependencies" might solve it (reverting #3477).

deivid-rodriguez commented:
Specifically, reverting https://github.com/decidim/decidim/commit/434e93c777e518841d4bb7375e418d23ad385faf.

mrcasals commented:
@deivid-rodriguez should we revert it from `master` then? Or only in this branch?

deivid-rodriguez commented:
If it works, I guess we should revert it from master as well. But I'm not sure it will work...

mrcasals commented:
@deivid-rodriguez can you check #3794 to see if the revert makes sense? I had some conflicts on the `Gemfile.lock` files...

mrcasals commented:
@deivid-rodriguez exactly, we have other backports failing (see #3789). I found an error on the cherry-pick, I've updated that other PR to see if tests pass correctly.

mrcasals commented:
I'm trying importing #3794 here, let's see if this fixes the tests...

mrcasals commented:
@deivid-rodriguez I seem to keep having an error with the tests when testing the `--edge` flag, as it installs decidim from GitHub but this raises errors:

https://circleci.com/gh/decidim/decidim/126279

Any idea on what's going on here? 😭

deivid-rodriguez commented:
No... It's the same issue as before, and I don't know why it happens. I think `bundler` should be able to resolve it, but it can't. So `bundler` bug I'd say. I guess we can skip the test and I'll try to follow up with the `bundler` team?

deivid-rodriguez commented:
Hehe, actually. If you look at the error message, it's using the "weird ~> a" constraint that we hypothesized could be causing the issue. That's because while we reverted it on this branch, it's not reverted on master. So technically, if this is what's causing it, it would happen untill we revert it from master, right?

Maybe a way to check is:

* See if the "branch" spec is failing (disable `--fail-fast` temporarily).
* If it fails, change the spec to point to this branch instead of to master `--branch 0.13-stable` instead of `--branch master`. If it starts passing, then the weird constraint was indeed the culprit and we can remove it from master.

mrcasals commented:
@deivid-rodriguez oh, makes sense. I've removed the rest of CI jobs for this branch, and removed the `FAIL_FAST` option so it only runs all generators tests. If it fails, then I'll do the other check 😄

deivid-rodriguez commented:
Good news for now, the branch spec also failed... :crossed_fingers:

mrcasals commented:
@deivid-rodriguez I've pushed a new commit changing the `--branch` spec, let's see...

mrcasals commented:
OK, so that's the problem. The `--branch` tests is passing correctly, while the `--edge` flag fails. I'll revert https://github.com/decidim/decidim/commit/434e93c777e518841d4bb7375e418d23ad385faf on `master`.

deivid-rodriguez commented:
Nice! I'll try to isolate this thing and report to `bundler` too!

mrcasals commented:
Check #3806 for a PR reverting 434e93c777e518841d4bb7375e418d23ad385faf on `master`.

mrcasals commented:
@deivid-rodriguez I've re-scheduled the job, so in my mind the `--edge` flag should be working, but it isn't 😞 

https://circleci.com/gh/decidim/decidim/126424

It now uses the correct version of the decidim dependencies, but it still has some errors, so *maybe* that commit wasn't causing the error 😕

deivid-rodriguez commented:
OMG, this is a nightmare! :sob:

mrcasals commented:
Let's try importing #3684, as it fixes some generators issues, and maybe it fixes our issues here 😓

mrcasals commented:
What bugs me the most is that we get this error while running the tests:

```
.F....ERROR:  While executing gem ... (Gem::InstallError)
    gem "decidim-generators" is not installed
```

How come it's not installed in this case? On `master` the tests are green, so it has to be something else. If we specify the `--branch` option and use the branch of this PR, then the tests are green, so this branch is OK too. So it has to be something about the integration of those two branches. Maybe we're missing a `with_clean_env` call somewhere (I don't know how that works)?

Going down the stacktrace we can see this:

```
Resolving dependencies.....
       Bundler could not find compatible versions for gem "decidim-generators":
         In Gemfile:
           decidim was resolved to 0.14.0.dev, which depends on
             decidim-generators (= 0.14.0.dev)
     
       Could not find gem 'decidim-generators (= 0.14.0.dev)', which is required by gem
       'decidim', in any of the sources.
     
       Bundler could not find compatible versions for gem "rubocop":
         In Gemfile:
           decidim-dev was resolved to 0.14.0.dev, which depends on
             rubocop (~> 0.54.0)
     
           decidim-dev was resolved to 0.14.0.dev, which depends on
             erb_lint (~> 0.0.22) was resolved to 0.0.26, which depends on
               rubocop (~> 0.51)
     
           decidim-dev was resolved to 0.14.0.dev, which depends on
             rubocop-rspec (~> 1.21) was resolved to 1.26.0, which depends on
               rubocop (>= 0.53.0)
     
       Bundler could not find compatible versions for gem "sass":
         In Gemfile:
           decidim was resolved to 0.14.0.dev, which depends on
             decidim-core (= 0.14.0.dev) was resolved to 0.14.0.dev, which depends on
       foundation-rails (< 6.5, ~> 6.4) was resolved to 6.4.3.0, which depends
       on
                 sass (< 3.5, >= 3.3.0)
     
           decidim was resolved to 0.14.0.dev, which depends on
             decidim-core (= 0.14.0.dev) was resolved to 0.14.0.dev, which depends on
               sassc (>= 1.12.1, ~> 1.12) was resolved to 1.12.1, which depends on
                 sass (>= 3.3.0)
     
           decidim was resolved to 0.14.0.dev, which depends on
             decidim-core (= 0.14.0.dev) was resolved to 0.14.0.dev, which depends on
               sassc-rails (~> 1.3) was resolved to 1.3.0, which depends on
                 sass
```

For some reason, it fails to install these three gems:
- `decidim-generators`
- `rubocop`
- `sass`

In this branch, all gems are updated to their latest possible version. This branch is using, specifically:
- `decidim-generators`: `0.13.0.pre1`
- `rubocop`: `0.54.0`
- `sass`: `3.4.25`

`master` uses:
- `decidim-generators`: `0.14.0.dev`
- `rubocop`: `0.54.0`
- `sass`: `3.4.25`

So, at least in my mind, bundler should be able to match the correct versions and install them. 

I tried running the failing test in my machine, but it works as expected, so I guess I might have some gems installed that make the test pass, for some reason?

josepjaume commented:
@mrcasals @deivid-rodriguez haven't thought much about this, but maybe updating the docker image to the latest rubygems and bundler might help?

mrcasals commented:
Maybe we're missing #3730 here?

deivid-rodriguez commented:
> .F....ERROR:  While executing gem ... (Gem::InstallError)
    gem "decidim-generators" is not installed

Yeah, this part is suspicious indeed. And I don't think it happens on other branches, but I don't know the reason...

The other error you copied is completely nonsense to me, since I don't see any contradictions. It looks like a form of https://github.com/bundler/bundler/issues/6620, maybe.

> @mrcasals @deivid-rodriguez haven't thought much about this, but maybe updating the docker image to the latest rubygems and bundler might help?

The base image already uses latest rubygems and bundler, I think... :(

I can actually reproduce this locally, so I'll have a closer look since I'm really curious about this... My bet is that this is a bug in `bundler` so I guess we could skip the test on this branch to unblock us for now.

deivid-rodriguez commented:
I think I found the culprit, let me try cook something up!

deivid-rodriguez commented:
Bug in parser, I think... :rage:. Either upgrade rubocop to 0.58.1, or lock parser to 2.5.1.0.

mrcasals commented:
@deivid-rodriguez yup, here's the report: https://github.com/whitequark/parser/issues/519 I tried upgrading `rubocop`, but then running `bundle exec rubocop` from the main folder only runs the lint to 30 files for some reason, so I'd downgrading `parser` here and I'll deal with the correct `rubocop` upgrade on `master`.

deivid-rodriguez commented:
Yeah, I recall that issue when upgrading other repos. I fixed it by adding 

```YAML
inherit_mode:
  merge:
    - Include
```
if I recall correctly.

mrcasals commented:
@deivid-rodriguez this seems to fix it for me on `master`: https://github.com/rubocop-hq/rubocop/issues/4247#issuecomment-389364734
